source/v1/I2CDetail.cpp

```diff
@@ -0,0 +1,235 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2016 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mbed-drivers/v1/I2CDetail.hpp"
+
+#if DEVICE_I2C && DEVICE_I2C_ASYNCH
+
+#include "mbed-drivers/v1/I2C.hpp"
+#include "core-util/CriticalSectionLock.h"
+#include "minar/minar.h"
+
+namespace mbed_drivers {
+using namespace mbed;
+namespace v1 {
+namespace detail {
+// TODO: Increase the size of I2COwners to accept Resource Mangers for non-onchip I2C masters
+I2CResourceManager * I2COwners[MODULES_SIZE_I2C] = {nullptr};
+
+int I2CResourceManager::post_transaction(I2CTransaction *t) {
+    int rc;
+    if ((rc = validate_transaction(t))) {
```

Comment:
Placing and assignment inside a condition in this manner is likely to cause confusion in the future, primarily to the next person that comes along and needs to work in this code, it would be better to hoist the assignment above the 'if'